AWS IAM auth 'Content-Length' header causes validation error when sent as number instead of string or []strings

**Environment:**
<!-- The version can be retrieved with `vault version`. -->

* Vault Version: v0.9.1
* Operating System/Architecture: Ubuntu 16.04 LTS on EC2 instance

**Vault Config File:**
<!-- Configuration file used for the vault server. -->
(dev server)

**Expected Behavior:**
<!-- What should have happened? -->
Vault returns the IAM auth login response, as described in [the docs](https://www.vaultproject.io/docs/auth/aws.html)

**Actual Behavior:**
<!-- What actually happened? -->
Vault returned this error:
`Error parsing iam_request_headers: header "Content-Length" value "43" has type json.Number, not string or []interface`

**Steps to Reproduce:**
<!-- List the steps required to reproduce the issue. -->
Create request and sign it with aws4 npm module:

**Original signed request (aws4)**
```
{ service: 'sts',
  body: 'Action=GetCallerIdentity&Version=2011-06-15',
  headers: 
   { 'X-Vault-AWS-IAM-Server-ID': 'i-99999999999993',
     Host: 'sts.amazonaws.com',
     'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8',
     'Content-Length': 43,
     'X-Amz-Security-Token': 'FQoaaaaEML//////////wEaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaL7O0gU=',
     'X-Amz-Date': 20210401T171213Z',
     Authorization: 'AWS4-HMAC-SHA256 Credential=AAAAAAAAAAAAAA/20210401/us-east-1/sts/aws4_request, SignedHeaders=content-length;content-type;host;x-amz-date;x-amz-security-token;x-vault-aws-iam-server-id, Signature=8aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa0' },
  method: 'POST',
  hostname: 'sts.amazonaws.com',
  path: '/' }
```

**Formatted for vault**
```
{ method: 'POST',
  path: '/auth/aws/login',
  json: 
   { role: 'some-role',
     iam_http_request_method: 'POST',
     iam_request_url: 'aHRaaaaaaaaaaaaaaaaaaaaaaavbS8=',
     iam_request_body: 'QWaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaQ==',
     iam_request_headers: 'eyJYLAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAaAafQ==' },
  uri: 'http://vault.vpc/v1/auth/aws/login',
  headers: {} }
```

**Important Factoids:**
<!-- Describe any atypical environment setup, if any. -->
- I faked the data up there for obvious reasons, but I can guarantee the base64 text contains the correct (stringified) version of the json (with 'Content-Length': 43 - as a number not a string)
- Running the dev server
- Vault `aws` backend initialized
- Vault AWS role and user are setup and access key and secret are configured in aws auth: the role policy is copied from the aws auth documentation without the `assumeRole' permission (not needed)
- No header enforcement setup (the header in the request serves no purpose for now, but I don't think it influences this issue)
- Using the `request-promise-native` npm module for requests
- Using the `aws4` module for aws request signing. This module adds the Content-Length header before signing the request, but it does so correctly (as a number).
- Programming this as part of the `node-vault` module (I'm adding some functionalities), [here's the failing code](https://github.com/DaniGuardiola/node-vault/blob/7525dc5a335d70be21e3cfd77becfc4bf9e9c94a/src/aws-auth.js)

**Please help!**
This is blocking and urgent for my company :/ thanks in advance!

@jefferai master failed to build. I am on the branch you specified now, building, I'll let you know the result.

Some of the master build errors (`make dev`):
```
==> Checking that code complies with gofmt requirements...
==> Checking that build is using go version >= 1.9.1...
go generate 
==> Removing old directory...
==> Building...
Number of parallel builds: 1

-->     linux/amd64: github.com/hashicorp/vault

1 errors occurred:
--> linux/amd64 error: exit status 2
Stderr: # github.com/hashicorp/vault/vendor/github.com/hashicorp/vault-plugin-auth-gcp/plugin
vendor/github.com/hashicorp/vault-plugin-auth-gcp/plugin/backend.go:53:12: cannot use b.pathLoginRenew (type func(*logical.Request, *framework.FieldData) (*logical.Response, error)) as type framework.OperationFunc in field value
vendor/github.com/hashicorp/vault-plugin-auth-gcp/plugin/path_config.go:32:25: cannot use b.pathConfigRead (type func(*logical.Request, *framework.FieldData) (*logical.Response, error)) as type framework.OperationFunc in map value
vendor/github.com/hashicorp/vault-plugin-auth-gcp/plugin/path_config.go:33:27: cannot use b.pathConfigWrite (type func(*logical.Request, *framework.FieldData) (*logical.Response, error)) as type framework.OperationFunc in map value
vendor/github.com/hashicorp/vault-plugin-auth-gcp/plugin/path_login.go:43:27: cannot use b.pathLogin (type func(*logical.Request, *framework.FieldData) (*logical.Response, error)) as type framework.OperationFunc in map value
vendor/github.com/hashicorp/vault-plugin-auth-gcp/plugin/path_login.go:44:35: cannot use b.pathLogin (type func(*logical.Request, *framework.FieldData) (*logical.Response, error)) as type framework.OperationFunc in map value
vendor/github.com/hashicorp/vault-plugin-auth-gcp/plugin/path_login.go:103:66: not enough arguments in call to framework.LeaseExtend(role.TTL, role.MaxTTL, b.Backend.System())
	have (*logical.Request, *framework.FieldData)
	want (context.Context, *logical.Request, *framework.FieldData)
```

Build failed as well on the `aws-jsonnumber` branch:
```
==> Checking that code complies with gofmt requirements...
==> Checking that build is using go version >= 1.9.1...
go generate 
==> Removing old directory...
==> Building...
Number of parallel builds: 1

-->     linux/amd64: github.com/hashicorp/vault

1 errors occurred:
--> linux/amd64 error: exit status 1
Stderr: go build github.com/hashicorp/vault: /usr/lib/go-1.9/pkg/tool/linux_amd64/link: signal: killed

Makefile:28: recipe for target 'dev' failed
make: *** [dev] Error 1
```

I'm retrying just in case

--- a/builtin/credential/aws/path_login.go
+++ b/builtin/credential/aws/path_login.go
@@ -5,6 +5,7 @@ import (
 	"crypto/subtle"
 	"crypto/x509"
 	"encoding/base64"
+	"encoding/json"
 	"encoding/pem"
 	"encoding/xml"
 	"fmt"
@@ -1467,11 +1468,15 @@ func parseIamRequestHeaders(headersB64 string) (http.Header, error) {
 		switch typedValue := v.(type) {
 		case string:
 			headers.Add(k, typedValue)
+		case json.Number:
+			headers.Add(k, typedValue.String())
 		case []interface{}:
 			for _, individualVal := range typedValue {
 				switch possibleStrVal := individualVal.(type) {
 				case string:
 					headers.Add(k, possibleStrVal)
+				case json.Number:
+					headers.Add(k, possibleStrVal.String())
 				default:
 					return nil, fmt.Errorf("header %q contains value %q that has type %s, not string", k, individualVal, reflect.TypeOf(individualVal))
 				}

